scipy/interpolate/_aaa.py

```diff
@@ -0,0 +1,363 @@
+# Copyright (c) 2017, The Chancellor, Masters and Scholars of the University
+# of Oxford, and the Chebfun Developers. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#     * Redistributions of source code must retain the above copyright
+#       notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above copyright
+#       notice, this list of conditions and the following disclaimer in the
+#       documentation and/or other materials provided with the distribution.
+#     * Neither the name of the University of Oxford nor the names of its
+#       contributors may be used to endorse or promote products derived from
+#       this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+import numpy as np
+import scipy
+
+
+class AAA:
+    r"""
+    AAA real or complex rational approximation.
+
+    As described in [1]_, the AAA algorithm is greedy algorithm for approximation by
+    rational functions on a real or complex set of points. The rational approximation is
+    represented in a barycentric form from which the zeros, poles, and residues can be
+    computed. 
+
+    Parameters
+    ----------
+    f : array_like or callable
+        Function values at `z` or a function that will be called `f(z)`.
+    z : array_like
+        Values at which `f` is provided.
+    rtol : float, optional
+        Relative tolerance, defaults to 1e-13.
+    max_terms : int, optional
+        maximal number of terms in the barycentric representation, defaults to 100.
+
+    Attributes
+    ----------
+    support_points : array
+        Support points of the approximation.
+    data_values : array
+        Value of the approximation at the `support_points`.
+    weights : array
+        Weights of the barycentric approximation.
+    errors : array
+        Error in the successive iterations of AAA.
+    poles : array
+        Poles of the AAA approximation.
+    residues : array
+        Residues associated with the poles of the approximation.
+    zeros : array
```

Comment:
I wonder if we should use `roots` instead of `zeros`. I think that terminology is more common in SciPy and doesn't get confused with `np.zeros` for instance.

Reply:
Good point will rename to roots (also called roots in [RationalFunctionApproximation.jl](https://complexvariables.github.io/RationalFunctionApproximation.jl/stable/))